cpp/src/parquet/encoding.cc

```diff
@@ -2466,6 +2468,99 @@ class DeltaByteArrayDecoder : public DecoderImpl,
 // ----------------------------------------------------------------------
 // BYTE_STREAM_SPLIT
 
+namespace {
+template <typename T, bool has_sse2_or_higher>
+struct ByteStreamSplitDecodePath {
+  static int Decode(T* buffer, int max_values, int num_values_in_buffer,
+                    const uint8_t* data, int& num_values_decoded, int& len_remaining) {
+    throw ParquetException("This should be unreachable.");
+  }
+};
+
+template <typename T>
+struct ByteStreamSplitDecodePath<T, true> {
```

Comment:
Can you implement some basic benchmarks in encoding_benchmark.cc for this?

Reply:
+1

Reply:
Yes. Done.

Reply:
So, the results on my Sandy Bridge 3930K is:
BM_ByteStreamSplitDecode_Float_Scalar/1024         1410 ns         1410 ns       486193 bytes_per_second=2.70593G/s
BM_ByteStreamSplitDecode_Float_Scalar/4096         5531 ns         5531 ns       123694 bytes_per_second=2.75865G/s
BM_ByteStreamSplitDecode_Float_Scalar/32768       44666 ns        44664 ns        11681 bytes_per_second=2.73305G/s
BM_ByteStreamSplitDecode_Float_Scalar/65536       88438 ns        88437 ns         6791 bytes_per_second=2.76063G/s
BM_ByteStreamSplitDecode_Float_SSE2/1024            178 ns          178 ns      3857270 bytes_per_second=21.483G/s
BM_ByteStreamSplitDecode_Float_SSE2/4096            723 ns          723 ns       968435 bytes_per_second=21.1157G/s
BM_ByteStreamSplitDecode_Float_SSE2/32768          6463 ns         6463 ns       104508 bytes_per_second=18.888G/s
BM_ByteStreamSplitDecode_Float_SSE2/65536         17350 ns        17349 ns        40253 bytes_per_second=14.072G/s
BM_ByteStreamSplitDecode_Double_Scalar/1024        3891 ns         3891 ns       181784 bytes_per_second=1.96083G/s
BM_ByteStreamSplitDecode_Double_Scalar/4096       15477 ns        15476 ns        45396 bytes_per_second=1.97188G/s
BM_ByteStreamSplitDecode_Double_Scalar/32768     123854 ns       123853 ns         5530 bytes_per_second=1.97122G/s
BM_ByteStreamSplitDecode_Double_Scalar/65536     256855 ns       256844 ns         2653 bytes_per_second=1.90108G/s
BM_ByteStreamSplitDecode_Double_SSE2/1024           413 ns          413 ns      1683687 bytes_per_second=18.4656G/s
BM_ByteStreamSplitDecode_Double_SSE2/4096          1800 ns         1800 ns       381460 bytes_per_second=16.9515G/s
BM_ByteStreamSplitDecode_Double_SSE2/32768        19688 ns        19688 ns        35003 bytes_per_second=12.4006G/s
BM_ByteStreamSplitDecode_Double_SSE2/65536        44692 ns        44691 ns        15740 bytes_per_second=10.9258G/s

The higher through for the SSE2 with small input than with large input is as expected. There are so few elements that it fits in the L1 cache.
1024: 1024 * 8 bytes = 8KiB input and 8KiB output < 32KiB
4096: 64 KiB work set, so this already spills to the L2 cache.
32768 and 65536 should fit in the L3 cache.

It might be that the SSE2 implementation could be made slightly faster and I haven't profiled it. I recall from some other experiments that I could get around 10 GiB/s on a single core on my system when data is always loaded from memory.

The scalar version seems to be bound by operations being issued.

Reply:
Impressive. This is much better than I imagined.